### ISSUE TITLE
Supporting the combination of SPMD and ZeRO.

### ISSUE THREAD

**Issue by baoleai** [pytorch/xla#5725]
## 🚀 Feature
Supporting the combination of SPMD and ZeRO.

## Motivation

The [ZeRO](https://arxiv.org/pdf/1910.02054.pdf) is an important method for saving memory in the Data Parallel (DP) scenario, and frameworks like Megatron offer combinations of DP, TP(Tensor Parallel), and [Distributed Optimizer](https://github.com/NVIDIA/Megatron-LM#distributed-optimizer) (ZeRO stage 1). While torch_xla currently supports FSDP (ZeRO stage 3) and SPMD individually, there are no existing examples of combining SPMD and ZeRO stage 1 or stage 2 at present. When training LLM using SPMD for model and data parallelism, it becomes essential to utilize ZeRO stage1/stage2 for memory optimization.

**Comment by JackCaoG** [comment]
@alanwaketan and @yeounoh can comment but we can't combine user FSDP(manually adding collective) and GSPMD right now and there is no obvious way of doing that. There are a couple technical difficulty including that in user mode FSDP we started one process per device and for GSPMD we always have one process for GSPMD per host.

What you can do is to use GSPMD to represent FSDP style sharding.

**Comment by alanwaketan** [comment]
SPMD 2D sharing is what combined ZeRO and megatron. Maybe that's what you are looking for?  https://pytorch2023.sched.com/event/1QBUk/lightning-talk-large-scale-distributed-training-with-dynamo-and-pytorchxla-spmd-yeounoh-chung-jiewen-tan-google

**Comment by baoleai** [comment]
Thank you for your suggestions. The feature I need is the combination of SPMD with ZeRO stage2 (similar to Megatron's Distributed Optimizer), which means further splitting gradients and optimizer state based on the rank of DP on top of DP and TP. 
With SPMD 2D, we can achieve the combination of FSDP (ZeRO stage 3) and TP. However, FSDP not only splits gradients and optimizer state but also splits weights. Perhaps I can add sharding annotation in the optimizer state to achieve the combination of ZeRO stage 2 and DP&TP.

**Comment by alanwaketan** [comment]
> Thank you for your suggestions @alanwaketan @JackCaoG. The feature I need is the combination of SPMD with ZeRO stage2 (similar to Megatron's Distributed Optimizer), which means further splitting gradients and optimizer state across DP ranks on top of DP and TP.
> 
> For "SPMD 2D sharding is what combined ZeRO and megatron", do you mean sharding weight into TP_NUM shards by a certain dimension and into FSDP_NUM shards by another dimension?
> 
> FSDP is ZeRO stage 3 which not only splits gradients and optimizer states but also splits weights. One possible approach to achieve a combination of ZeRO stage 2 (which shards gradients and optimizer states across DP ranks) with DP&TP could be to add sharding annotations on the optimizer states instead of weights？

Can you send me more readings? Maybe we can discuss this in a meeting to see how we can achieve a similar effects just using SPMD annotation. Are you on slack?

**Comment by baoleai** [comment]
Hi @alanwaketan , I have already sent the request to join the Slack : )

**Comment by alanwaketan** [comment]
@baoleai So are you able to implement this sharding yourself?

**Comment by baoleai** [comment]
Yes, by sharding the gradients, it is possible to achieve ZeRO2.

**Comment by alanwaketan** [comment]
> Yes, by sharding the gradients, it is possible to achieve ZeRO2.

Would love to learn more about it, if you don't mind sharing some of the code examples?
